### ISSUE TITLE
Ask a question about GRANDE

### ISSUE THREAD

**Issue by ZZC2002ZZC** [s-marton/GRANDE#1]
Dear author,
I have read your paper on GRANDE and found it beneficial to my work. The method you proposed is used to process tabular data, so I want to use it to process oil well data for lithology inversion, but I have encountered a problem.
The well data includes 5 real data indicators, all of which I aim to use to predict the lithology. The lithology is divided into 12 classes, which are numbered by integers [30000,65030,65000,80000,74000,70000,70032,88000,86000,99000,90000,93000]. However, when I used your code to predict lithology, there was a mistake of "classes should include all valid labels that can be in y". I would like to ask you how to modify the code to apply to my current situation.
I would appreciate it very much if you could give me some suggestions during your busy schedule!The cleaned data is attached.

[cleaned_data1.xlsx](https://github.com/s-marton/GRANDE/files/14446460/cleaned_data1.xlsx)

**Comment by Tsesterh** [comment]
Hi,
unfortunately I could not replicate your issue, but I think this could arise due to the encoding of the lithology classes. Therefore, I created a script that assigns each class to a unique integer starting with 0, so 
30000 -> 0
65000 -> 1
65030 -> 2
...

Hope this helps, here is the code (you have to adapt the number of epochs, estimators, etc.):

```
from GRANDE import GRANDE
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, accuracy_score


df = pd.read_excel('cleaned_data1.xlsx')

# Convert 'LITHOLOGY' column to categorical and then to its codes
df['LITHOLOGY'] = pd.Categorical(df['LITHOLOGY']).codes
print(df.head())

y = df['LITHOLOGY']
X = df.drop(columns=['LITHOLOGY'])

# Split the data
X_temp, X_test, y_temp, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_valid, y_train, y_valid = train_test_split(X_temp, y_temp, test_size=0.25, random_state=42)


params = {
        'depth': 5,
        'n_estimators': 10,

        'learning_rate_weights': 0.005,
        'learning_rate_index': 0.01,
        'learning_rate_values': 0.01,
        'learning_rate_leaf': 0.01,

        'optimizer': 'SWA',
        'cosine_decay_steps': 0,

        'initializer': 'RandomNormal',

        'loss': 'crossentropy',
        'focal_loss': False,
        'temperature': 0.0,

        'from_logits': True,
        'apply_class_balancing': True,

        'dropout': 0.0,

        'selected_variables': 0.8,
        'data_subset_fraction': 1.0,
}

args = {
    'epochs': 1,
    'early_stopping_epochs': 25,
    'batch_size': 64,

    'cat_idx': [],
    'objective': 'binary',

    'metrics': ['F1'], # F1, Accuracy, R2
    'random_seed': 42,
    'verbose': 1,
}

model_grande = GRANDE(params=params, args=args)

model_grande.fit(X_train=X_train,
          y_train=y_train,
          X_val=X_valid,
          y_val=y_valid)


preds = model_grande.predict(X_test)

preds_classes = np.argmax(preds, axis=1)

accuracy = accuracy_score(y_test, preds_classes)
f1 = f1_score(y_test, preds_classes, average='macro')

print('Accuracy:', accuracy)
print('F1 Score:', f1)
```

**Comment by s-marton** [comment]
Hi,
currently GRANDE requires an ordinal encoding for (multi-class) classification tasks. This should be addressed by the code provided by @Tsesterh. 

@ZZC2002ZZC did this solve you issue? If there is any problem remaining, please let me know.

**Comment by ZZC2002ZZC** [comment]
Dear author,
Thank you and Tsesterh very much for your response to my question. Using the code provided by Tsesterh, I have solved the previous issue. 
However, when I used GRANDE for lithology(multi-class) classification, the accuracy was relatively low, and I did not know what caused it.
Would you please look at it if you have spare time. I have attached a small lithology classification data set and code results. In the code, I compared GRANDE with traditional randomforest, and the accuracy of GRANDE was less than 20%, while the accuracy of using randomforest was more than 80%.
Thank you again for your reply.
------------------ 原始邮件 ------------------
发件人:                                                                                                                        "s-marton/GRANDE"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年3月9日(星期六) 凌晨0:21
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [s-marton/GRANDE] Ask a question about GRANDE (Issue #1)





 
Hi,
 currently GRANDE requires an ordinal encoding for (multi-class) classification tasks. This should be addressed by the code provided by @Tsesterh.
 
@ZZC2002ZZC did this solve you issue? If there is any problem remaining, please let me know.

**Comment by s-marton** [comment]
This souds like an issue with the current implementation. I will have a look at it.

**Comment by s-marton** [comment]
@ZZC2002ZZC @Tsesterh 

After reviewing the code, I identified the issue as being related to the configuration of the 'objective' parameter within the args dictionary. It appears to be currently set to 'binary', which is suitable for binary classification tasks. However, for multi-class classification, this parameter should be changed to 'classification'. Retaining the 'binary' setting restricts GRANDE's predictions to a range between 0 and 1, effectively limiting the model to distinguish between only two out of twelve classes. This misconfiguration is likely the cause of the observed suboptimal performance.

Furthermore, I would like to draw attention to a particular hyperparameter setting in GRANDE that could have a significant impact on your specific scenario. Currently, GRANDE is configured to automatically incorporate class weights. This adjustment is designed to compensate for disparities in class representation by assigning greater importance to underrepresented classes. This feature is particularly relevant in scenarios where data imbalance is pronounced. For example, in the dataset subset you've provided, there are only 5 instances of the class '86000'. Although accurately predicting these underrepresented classes might be crucial for practical applications, it's important to note that their correct prediction has only a small impact on the accuracy metric. Therefore, you might consider a balanced accuracy of macro F1 score to account for class imbalance. To exclude class weights for GRANDE, you could remove line 195 (```class_weight = self.class_weight_dict,```) in ```GRANDE.py```.

When running the following code, GRANDE achieves an accuracy of 0.77 for GRANDE (0.80 when excluding class weights):

```
#specify GPU to use
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

from sklearn.model_selection import train_test_split
import openml
import category_encoders as ce
import numpy as np
import pandas as pd
import sklearn

df = pd.read_excel('cleaned_data1.xlsx')

# Convert 'LITHOLOGY' column to categorical and then to its codes
df['LITHOLOGY'] = pd.Categorical(df['LITHOLOGY']).codes

y = df['LITHOLOGY']
X = df.drop(columns=['LITHOLOGY'])

categorical_feature_indices = []

X_temp, X_test, y_temp, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train, X_valid, y_train, y_valid = train_test_split(X_temp, y_temp, test_size=0.2, random_state=42)

display(X_train.head())

print("Training set size:", len(X_train))
print("Validation set size:", len(X_valid))
print("Test set size:", len(X_test))

from GRANDE import GRANDE

params = {
        'depth': 5,
        'n_estimators': 2048,

        'learning_rate_weights': 0.005,
        'learning_rate_index': 0.01,
        'learning_rate_values': 0.01,
        'learning_rate_leaf': 0.01,

        'optimizer': 'SWA',
        'cosine_decay_steps': 0,

        'initializer': 'RandomNormal',

        'loss': 'crossentropy',
        'focal_loss': False,
        'temperature': 0.0,

        'from_logits': True,
        'apply_class_balancing': True,

        'dropout': 0.0,

        'selected_variables': 0.8,
        'data_subset_fraction': 1.0,
}

args = {
    'epochs': 1_000,
    'early_stopping_epochs': 25,
    'batch_size': 64,

    'cat_idx': categorical_feature_indices,
    'objective': 'classification',
    
    'metrics': ['F1'], # F1, Accuracy, R2
    'random_seed': 42,
    'verbose': 1,       
}

model_grande = GRANDE(params=params, args=args)

model_grande.fit(X_train=X_train,
          y_train=y_train,
          X_val=X_valid,
          y_val=y_valid)

preds_grande = model_grande.predict(X_test)
```

However, I want to note that such an extreme imbalance (only 5/239,478 instances for one class) is hard to capture for GRANDE (e.g. due to the mini-batch training) and therefore might require some modifications to the architecture/hyperparameters and/or the training procedure. 

Let me know if you face further issues. :-)
